### PR TITLE
rec: First check the cache for NS name to address contents for both v4 and v6

### LINKED ISSUE #10263: pdns-recursor fails to resolve records for domain if qname-minimization is enabled
- Program: Recursor
 - Issue type: Bug report

### Short description
pdns-recursor versions 4.4.0 onwards fail to resolve records from a specific domain, queries return SERVFAIL. 
The same queries work fine in 4.3.7. 

### Environment
 - Operating system: Alpine Linux 3.13
 - Software version: 4.3.7 (successful) / 4.4.0+ (fails)
 - Software source: Source tarballs, self compiled
 - Configure options: ./configure --enable-static --enable-static-boost --with-protobuf
 
### Steps to reproduce

The problem domain consists of 3, IPv6 only, NS records, 

```text
welterde.dn42.  3502    IN      SOA     ns1.any.welterde.dn42. hostmaster.welterde.dn42. 2021040520 28800 3600 2419200 3600
welterde.dn42.  3600    IN      NS      ns1.any.welterde.dn42.
welterde.dn42.  3600    IN      NS      ns2.any.welterde.dn42.
welterde.dn42.  3600    IN      NS      ns3.any.welterde.dn42.
ns1.any.welterde.dn42.  3600    IN      AAAA    fdef:5d0d:ee12:0:100:53:42:1
ns2.any.welterde.dn42.  3600    IN      AAAA    fdef:5d0d:ee12:0:100:53:42:2
ns3.any.welterde.dn42.  3600    IN      AAAA    fdef:5d0d:ee12:0:100:53:42:3
```

The .dn42 TLD is forwarded by the recursor to an authoritative server (pdns) using a forward-zones files:

```text
dn42 = [fd42:4242:2601:ac53::1]:53
```

A single delegation and glue for ns1.any.welterde.dn42 exists in the .dn42 zone:

```text
welterde.dn42   900     IN      NS      ns1.any.welterde.dn42.
ns1.any.welterde.dn42   900     IN      AAAA    fdef:5d0d:ee12:0:100:53:42:1
```

Querying the records using pdns-recursor 4.3.7 is successful:

```shell
# pdns_recursor --version
Apr 05 19:16:44 Unable to open /usr/local/etc/recursor.conf
Apr 05 19:16:44 Unable to parse configuration file '/usr/local/etc/recursor.conf'
Apr 05 19:16:44 Asked to run with pdns-distributes-queries set but no distributor threads
, raising to 1
Apr 05 19:16:44 PowerDNS Recursor 4.3.7 (C) 2001-2020 PowerDNS.COM BV
Apr 05 19:16:44 Using 64-bits mode. Built using gcc 10.2.1 20201203 on Apr  5 2021 18:33:
27 by simon@abuild.
Apr 05 19:16:44 PowerDNS comes with ABSOLUTELY NO WARRANTY. This is free software, and yo
u are welcome to redistribute it according to the terms of the GPL version 2.
Apr 05 19:16:44 Features: fcontext libcrypto-ecdsa libcrypto-ed25519 libcrypto-ed448 libc
rypto-eddsa lua nod protobuf sodium 
Apr 05 19:16:44 Configured with: " '--enable-static' '--enable-static-boost' '--with-prot
obuf'"

# drill -p 5353 AAAA ns2.any.welterde.dn42 @fd42:4242:2601:3502:216:3eff:fe11:fdbd
;; ->>HEADER<<- opcode: QUERY, rcode: NOERROR, id: 44383
;; flags: qr rd ra ; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0 
;; QUESTION SECTION:
;; ns2.any.welterde.dn42.       IN      AAAA

;; ANSWER SECTION:
ns2.any.welterde.dn42.  3600    IN      AAAA    fdef:5d0d:ee12:0:100:53:42:2

;; AUTHORITY SECTION:

;; ADDITIONAL SECTION:

;; Query time: 69 msec
;; SERVER: fd42:4242:2601:3502:216:3eff:fe11:fdbd
;; WHEN: Mon Apr  5 19:16:50 2021
;; MSG SIZE  rcvd: 67
```

Performing the same query on pdns-recursor version 4.4.0 and later results in a SERVFAIL
The trace for this query appears to show the recursor looping on the same queries until it eventually fails.

```sh
# pdns_recursor --version
Apr 05 19:09:49 PowerDNS Recursor 4.4.0 (C) 2001-2020 PowerDNS.COM BV
Apr 05 19:09:49 Using 64-bits mode. Built using gcc 10.2.1 20201203 on Apr  5 2021 18:40:
45 by simon@abuild.
Apr 05 19:09:49 PowerDNS comes with ABSOLUTELY NO WARRANTY. This is free software, and yo
u are welcome to redistribute it according to the terms of the GPL version 2.
Apr 05 19:09:49 Features: fcontext libcrypto-ecdsa libcrypto-ed25519 libcrypto-ed448 libc
rypto-eddsa lua nod protobuf sodium 
Apr 05 19:09:49 Configured with: " '--enable-static' '--enable-static-boost' '--with-prot
obuf'"

# drill -p 5353 AAAA ns2.any.welterde.dn42 @fd42:4242:2601:3502:216:3eff:fe11:fdbd
;; ->>HEADER<<- opcode: QUERY, rcode: SERVFAIL, id: 49367
;; flags: qr rd ra ; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 0 
;; QUESTION SECTION:
;; ns2.any.welterde.dn42.       IN      AAAA

;; ANSWER SECTION:

;; AUTHORITY SECTION:

;; ADDITIONAL SECTION:

;; Query time: 76 msec
;; SERVER: fd42:4242:2601:3502:216:3eff:fe11:fdbd
;; WHEN: Mon Apr  5 19:09:59 2021
;; MSG SIZE  rcvd: 39
```

Attached are the server config and logs with trace enabled for both 4.3.7 (successful) and 4.4.0 (failure). The traces show a complete log from clean restart to end of the query. 
The server would normally run with DNSSEC enabled, however this has been disabled for the purposes of the test, The server is also configured to run on port 5353 for testing. 

[recursor-dns53.conf.txt](https://github.com/PowerDNS/pdns/files/6260398/recursor-dns53.conf.txt)
[trace-4.4.0.txt](https://github.com/PowerDNS/pdns/files/6260399/trace-4.4.0.txt)
[trace-4.3.7.txt](https://github.com/PowerDNS/pdns/files/6260402/trace-4.3.7.txt)

### Expected behaviour
The domain records should resolve successfully in 4.4.0 and later.

### Actual behaviour
Records fail to resolve.

### Other information
none

### PR DESCRIPTION
**Pull request by omoerbeek**
This fixes #10263 here, but needs thorough reviewing and testing.

### Checklist

I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] compiled this code
- [X] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: pdns/syncres.cc [hunk 1]
```diff
@@ -1074,40 +1074,64 @@ vector<ComboAddress> SyncRes::getAddrs(const DNSName &qname, unsigned int depth,
...
-    res_t resv4;
-    // If IPv4 ever becomes second class, we should revisit this
-    if (s_doIPv4 && doResolve(qname, QType::A, resv4, depth+1, beenthere, newState) == 0) {  // this consults cache, OR goes out
-      for (auto const &i : resv4) {
-        if (i.d_type == QType::A) {
+    // First look for both A and AAAA in the cache and be satisfied if we find anything
+    res_t cset;
+    if (s_doIPv4 && g_recCache->get(d_now.tv_sec, qname, QType::A, false, &cset, d_cacheRemote, d_refresh, d_routingTag) > 0) {
+      for (const auto &i : cset) {
+        if (i.d_ttl > (unsigned int)d_now.tv_sec ) {
```
Comment (rgacogne) [inline pdns/syncres.cc:1081]: I don't think we need that test, `MemRecursorCache::get` should only return a positive value is the TTD is still good?

**Review by rgacogne — COMMENTED** [review]
The general idea makes sense to me and the code looks good.

File: pdns/syncres.cc (hunk 1 above)
Comment (omoerbeek) [reply pdns/syncres.cc:1081]: Right, this was a c/p from the existing cide. I'll double check and adapt.

Comment (omoerbeek) [reply pdns/syncres.cc:1081]: The return valiue of `get()` reflects the ttd of the cache entry, which should be <= the minimum of the ttd of the individual records in the cache entry. So I think the condition can be removed indeed.

File: pdns/syncres.cc [hunk 2]
```diff
@@ -1074,40 +1074,60 @@ vector<ComboAddress> SyncRes::getAddrs(const DNSName &qname, unsigned int depth,
...
+            }
+          }
+        } else {
+          // We have some IPv4 records, don't bother with going out to get IPv6, but do consult the cache, we might have
+          // encountered some v6 glue
+          // Once IPv6 adoption matters, this needs to be revisited
+          cset.clear();
+          if (g_recCache->get(d_now.tv_sec, qname, QType::AAAA, false, &cset, d_cacheRemote, d_refresh, d_routingTag) > 0) {
+            for (const auto &i : cset) {
+              if (i.d_ttl > (unsigned int)d_now.tv_sec ) {
```
Comment (rgacogne) [inline pdns/syncres.cc:1127]: I think the TTL check can go here as well, sorry :-/

File: pdns/syncres.cc [hunk 3]
```diff
@@ -1074,40 +1074,60 @@ vector<ComboAddress> SyncRes::getAddrs(const DNSName &qname, unsigned int depth,
...
+        if (auto rec = getRR<AAAARecordContent>(i)) {
+          ret.push_back(rec->getCA(53));
+        }
+      }
+    }
+    if (ret.empty()) {
+      // Nothing in the cache...
+      vState newState = vState::Indeterminate;
+      cset.clear();
+      // If IPv4 ever becomes second class, we should revisit this
```
Comment (rgacogne) [inline pdns/syncres.cc:1097]: Perhaps this comment would make more sense moved to line 1108 now?

**Review by rgacogne — APPROVED** [review]
LGTM!

**Pull request merged**